examples/ucp_client_server.c

```diff
@@ -95,6 +114,146 @@ static struct {
  */
 static void usage(void);
 
+void buffer_free(data_meta_t *mdata)
+{
+    size_t dt_iov_idx;
+    ucp_dt_iov_t *dt_iov;
+
+    if (mdata->send_recv_type == CLIENT_SERVER_SEND_RECV_TAG &&
+        mdata->data_type == DATATYPE_IOV) {
+        dt_iov = mdata->buffer;
+
+        for (dt_iov_idx = 0; dt_iov_idx < mdata->iov_num; dt_iov_idx++) {
+            if (dt_iov[dt_iov_idx].buffer == NULL) {
+                continue;
+            }
+            mem_type_free(dt_iov[dt_iov_idx].buffer);
+            dt_iov[dt_iov_idx].buffer = NULL;
+            dt_iov[dt_iov_idx].length = 0;
+        }
+    }
+    mem_type_free(mdata->buffer);
+    mdata->buffer = NULL;
+}
+
+int buffer_malloc(data_meta_t *mdata)
+{
+    ucp_dt_iov_t *dt_iov;
+    size_t dt_iov_idx;
+
+    if (mdata->send_recv_type == CLIENT_SERVER_SEND_RECV_TAG &&
+        mdata->data_type == DATATYPE_IOV) {
+        mdata->buffer = calloc(mdata->iov_num, sizeof(ucp_dt_iov_t));
+        dt_iov        = mdata->buffer;
+        CHKERR_ACTION(dt_iov == NULL, "allocate memory\n", return -1;);
+
+        for (dt_iov_idx = 0; dt_iov_idx < mdata->iov_num; dt_iov_idx++) {
+            dt_iov[dt_iov_idx].length = mdata->iov_sizes[dt_iov_idx];
+            dt_iov[dt_iov_idx].buffer = mem_type_malloc(
+                    dt_iov[dt_iov_idx].length);
+            if (dt_iov[dt_iov_idx].buffer == NULL) {
+                buffer_free(mdata);
```

Comment:
In this error case we still return 0? why not -1?

Reply:
Done